### PR TITLE
Add configurable `val_period` for periodic validation during training

### PR DESCRIPTION
**Pull request by Bovey0809**
Adds `val_period` parameter (default 1) to control how often validation runs during training. On non-validation epochs, CSV writes empty cells for val metrics (parsed as NaN) instead of stale values.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🚀 Adds configurable periodic validation during training with clean metric logging for epochs where validation is skipped.

### 📊 Key Changes

- Introduces the `val_period` setting, defaulting to `1` to validate every epoch.
- Validates only every N epochs when `val=True`, while still validating on the final epoch or when early stopping requires it.
- Records empty validation metric fields on skipped epochs instead of reusing stale values.
- Updates CSV metric formatting to preserve empty values while formatting numeric values normally.
- Adds configuration validation to ensure `val_period` is an integer of at least `1`.

### 🎯 Purpose & Impact

- ⚡ Reduces training overhead by allowing users to validate less frequently, such as with `val_period=5`.
- 📈 Produces more accurate training logs by clearly distinguishing skipped validation epochs from measured ones.
- 🛑 Preserves final-epoch and early-stopping validation behavior, helping maintain reliable training completion and stopping decisions.
- 🔧 Existing workflows are unchanged because the default still validates every epoch.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an Ultralytics engineer will assist shortly. ✅

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -513,7 +513,14 @@ def _do_train(self):
 
             # Validation
             final_epoch = epoch + 1 >= self.epochs
-            if self.args.val or final_epoch or self.stopper.possible_stop or self.stop:
+            val_period = max(self.args.val_period, 1)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `should_val` is gated by `self.args.val`, which means final-epoch and early-stop validation will no longer run when `val=False`. The previous behavior validated on `final_epoch`/`possible_stop`/`stop` regardless of `val`. This changes semantics and breaks the stated intent. Consider separating periodic validation from the forced final/early-stop validation.

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -523,8 +530,10 @@ def _do_train(self):
 
             self.nan_recovery_attempts = 0
             if RANK in {-1, 0}:
-                self.save_metrics(metrics={**self.label_loss_items(self.tloss), **self.metrics, **self.lr})
-                self.stop |= self.stopper(epoch + 1, self.fitness) or final_epoch
+                # On non-val epochs, write empty val metrics so CSV doesn't carry stale values
+                val_metrics = self.metrics if should_val else {k: "" for k in self.metrics}
+                self.save_metrics(metrics={**self.label_loss_items(self.tloss), **val_metrics, **self.lr})
+                self.stop |= self.stopper(epoch + 1, self.fitness if should_val else None) or final_epoch
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: Passing `None` into `self.stopper(...)` on non-validation epochs can raise if the stopper expects a float and may disable early-stopping logic unexpectedly. It’s safer to pass the last known `self.fitness` (or skip calling the stopper) rather than `None`.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good addition of `val_period` and CSV handling, but there are two logic issues in the training loop: forced validation on final/early-stop is now suppressed when `val=False`, and `self.stopper` receives `None` on skipped-val epochs which can break or alter early-stopping behavior. Fixing these will align behavior with the PR’s stated intent.

💬 Posted 2 inline comments
